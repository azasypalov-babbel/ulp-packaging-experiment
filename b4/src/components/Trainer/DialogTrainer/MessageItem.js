import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { rem } from 'polished';
import { markupStringToHTML } from '@lessonnine/babbel-markup-helper.js';

import Text from '../../shared/Text';

const activeStyles = css`
  background-color: ${(props) => props.theme.color.pastel.electric};
  border-color: ${(props) => props.theme.color.pastel.electric};
`;

const leftStyles = css`
  border-radius: 0 ${rem('12px')} ${rem('12px')} ${rem('12px')};
`;

const rightStyles = css`
  border-radius: ${rem('12px')} 0 ${rem('12px')} ${rem('12px')};
`;

const StyledWrapper = styled.div`
  box-sizing: border-box;
  display: inline-flex;
  flex-direction: column;

  max-width: 100%;

  padding: ${rem('16px')};

  border: 2px solid ${(props) => props.theme.color.silver};
  background-color: ${(props) => props.theme.color.semantic.interactive.secondary.background};

  transition: border 0.1s linear, background-color 0.1s linear;

  & > * + * {
    margin-top: ${rem('8px')};
  }

  &:hover {
    border: 2px solid ${(props) => props.theme.color.semantic.interactive.secondary.border};

    ${({ active }) => active && activeStyles}
  }

  ${({ active }) => active && activeStyles}

  ${({ position }) => {
    if (position === 'left') return leftStyles;
    if (position === 'right') return rightStyles;
  }};
`;

const MessageItem = (props) => {
  const {
    children,
    secondaryText,
    active,
    position,
    onClick
  } = props;
  return (
    <StyledWrapper
      active={active}
      position={position}
      onClick={onClick}
    >
      <Text
        fontFamily="fontMilliardBook"
        color="spaceGray"
      >
        {children}
      </Text>
      { secondaryText &&
        <Text
          data-selector="item-translation"
          fontFamily="fontMilliardBook"
          color="spaceGrayW28"
          fontSize="small"
          dangerouslySetInnerHTML={{ __html: markupStringToHTML(secondaryText) }}
        />
      }
    </StyledWrapper>
  );
};

MessageItem.propTypes = {
  children: PropTypes.node.isRequired,
  secondaryText: PropTypes.string,
  active: PropTypes.bool,
  onClick: PropTypes.func,
  position: PropTypes.oneOf(['left', 'right']).isRequired
};

export default MessageItem;
