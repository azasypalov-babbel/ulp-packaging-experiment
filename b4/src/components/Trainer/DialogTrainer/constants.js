export const TIMEOUT_AFTER_CONTEXTUAL_INFO = 2000;
