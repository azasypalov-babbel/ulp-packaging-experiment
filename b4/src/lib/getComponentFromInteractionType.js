import Fillin from '../components/Interactions/Fillin/FillinInteraction';
import Puzzlehelper from '../components/Interactions/Puzzlehelper/PuzzlehelperInteraction';
import Wordorder from '../components/Interactions/Wordorder/WordorderInteraction';
import ChoiceButtons from '../components/Interactions/Choicebuttons/ChoicebuttonsInteraction';
import Speak from '../components/Interactions/Speak/SpeakInteraction';

const interactionMap = {
  speak: Speak,
  fillin: Fillin,
  puzzlehelper: Puzzlehelper,
  wordorder: Wordorder,
  choose: ChoiceButtons
};

const getComponentFromInteractionType = (interactionType) => interactionMap[interactionType];

export default getComponentFromInteractionType;

