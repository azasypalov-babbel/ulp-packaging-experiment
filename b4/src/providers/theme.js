import React, { useContext } from 'react';
import { ThemeProvider as StyledThemeProvider, ThemeContext } from 'styled-components';
import { getDisplayName } from '../components/getDisplayName';

import tokens from '@lessonnine/design-tokens.lib/dist/json/variables';

export const useTheme = () => useContext(ThemeContext);

const size = {
  font: {
    xsmall: '0.75rem',
    small: '0.875rem',
    base: '1rem',
    medium: '1.125rem',
    big: '1.5rem',
    large: '2rem',
    xlarge: '2.5rem',
    grand: '3rem'
  },
  navbar: {
    xsmall: {
      height: '60px' // Approximate height of navigation bar in `xsmall` environments
    }
  }
};

export const theme = {
  cascada: tokens.legacy.cascada,
  color: tokens.color,
  viewports: tokens.viewports,
  size
};

const withThemeProvider = (WrappedComponent) => {
  const ThemeProvider = (props) => (
    <StyledThemeProvider theme={theme}>
      <WrappedComponent {...props} />
    </StyledThemeProvider>
  );

  ThemeProvider.displayName = `ThemeProvider(${getDisplayName(WrappedComponent)})`;

  return ThemeProvider;
};

export default withThemeProvider;
