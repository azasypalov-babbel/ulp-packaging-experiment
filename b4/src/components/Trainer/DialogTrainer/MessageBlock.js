import React, { useEffect, useRef, useState } from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import SpeakerAvatar from './SpeakerAvatar';
import MessageItem from './MessageItem';
import { rem } from 'polished';
import { scrollIntoSafeArea } from '../../shared/scrollIntoSafeArea';


const StyledMessageBlock = styled.div`
  display: flex;
  flex-direction: column;

  cursor: ${(props) => props.onClick ? 'pointer' : 'initial'};

  ${(props) => props.hasAvatar && css`margin-top: ${rem('12px')};`};

  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    flex-direction: row;
  }

  & > div:first-child {
    margin: 0 0 ${rem('12px')} 0;
    ${({ position }) => {
    if (position === 'right') return css`
        @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
          margin: 0 0 0 ${rem('12px')};
        }
      `;

    if (position === 'left') return css`
      @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
        margin: 0 ${rem('12px')} 0 0; }`;
  }}
  }
`;

const StyledRow = styled.div`
  display: flex;
  align-items: flex-start;

  &:focus {
    outline: none;
  }

  ${({ position }) => {
    if (position === 'left') return css`
      flex-direction: row;
    `;
    if (position === 'right') return css`
      flex-direction: row-reverse;
      & > div {
        flex-direction: column;
        align-items: flex-end;
      }

      @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
        & > div {
          align-items: flex-start;
          flex-direction: row-reverse;
        }
      }
    `;
  }};
`;

const AvatarSpacer = styled.div`
  height: ${rem('48px')};
  width: ${rem('48px')};
  min-width: ${rem('48px')};
  display: none;
  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xsmall}) {
    display: block;
  }
`;

const MessageBlock = (props) => {
  const {
    hasAvatar,
    speakerRole,
    children,
    secondaryText,
    position,
    isSpeaking,
    onClick,
    avatarIsSpeaking,
    active,
    dataItemType
  } = props;

  const [isSpeakingByClick, setIsSpeakingByClick] = useState(false);

  const handleClick = (event) => {
    setIsSpeakingByClick(true);
    onClick(event);
  };

  useEffect(() => {
    // This a stateful variant of the isSpeaking prop.
    // It will only be true if isSpeaking was triggered by a user click.
    if (isSpeakingByClick && !isSpeaking) {
      setIsSpeakingByClick(false);
    }
  }, [isSpeakingByClick, isSpeaking]);

  const ref = useRef();

  useEffect(() => {
    if (active) {
      scrollIntoSafeArea(ref.current);
    }
  }, [ref, active]);

  return (
    <StyledRow
      ref={ref}
      position={position}
      tabIndex="0"
    >
      <StyledMessageBlock
        active={isSpeakingByClick}
        onClick={onClick ? handleClick : undefined}
        position={position}
        data-selector="item-sound-replay-wrapper"
        hasAvatar={hasAvatar}
      >
        {hasAvatar ? <SpeakerAvatar
          speakerRole={speakerRole}
          isSpeaking={avatarIsSpeaking}
        /> : <AvatarSpacer />}
        <div position={position} data-item-type={dataItemType}>
          <MessageItem
            secondaryText={secondaryText}
            active={isSpeakingByClick}
            position={position}
          >
            {children}
          </MessageItem>
        </div>
      </StyledMessageBlock>
    </StyledRow>
  );
};

MessageBlock.propTypes = {
  position: PropTypes.oneOf(['left', 'right']).isRequired,
  active: PropTypes.bool,
  avatarIsSpeaking: PropTypes.bool,
  isSpeaking: PropTypes.bool,
  hasAvatar: PropTypes.bool,
  children: PropTypes.node,
  speakerRole: PropTypes.oneOf(['m1', 'm2', 'f1', 'f2', 'n1', 'None']),
  secondaryText: PropTypes.string,
  onClick: PropTypes.func,
  dataItemType: PropTypes.string
};

export default MessageBlock;
