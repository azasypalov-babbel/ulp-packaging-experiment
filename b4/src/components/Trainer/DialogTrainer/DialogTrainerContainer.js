import React, { useState, useEffect, useMemo, useCallback, useRef } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import * as sequenceActions from '../../../dux/sequence/actions';
import withInfoText from '../../shared/withInfoText';
import { withServices } from '../../shared/withServices';
import ContinueSheet from '../../ContinueSheet';
import DialogTrainer from './DialogTrainer';
import { useSequence } from '../../shared/useSequence';
import withNormalisedTrainerData from '../withNormalisedTrainerData';
import flattenItems from '../shared/flattenItems';
import useTrainerItemSounds from '../shared/useTrainerItemSounds';
import { isInteractive } from '../../shared/itemSelectors';
import { TIMEOUT_AFTER_INFOTEXT_DISPLAY } from '../../shared/InfoText/constants';
import { TIMEOUT_AFTER_CONTEXTUAL_INFO } from './constants';
import { wait } from '../../shared/wait';

const stripHtml = (s) => s ? s.replace(/(<([^>]+)>)/ig, '') : '';

const TRAINER_STATE = {
  READY: 'READY',
  AWAITING_CONTINUE: 'AWAITING_CONTINUE',
  DONE: 'DONE'
};

const getInteractionType = (trainer) => {
  if (trainer.interaction === 'write') {
    return trainer.puzzleHelper ? 'puzzlehelper' : 'fillin';
  }
  return trainer.interaction;
};

const createDescriptionAsTask = (description) => ({
  id: '0',
  type: 'task',
  displayLanguageText: description
});

export const DialogTrainerContainer = ({
  attemptItem,
  completeItem,
  currentTrainerItemIndex: initialItemIndex,
  displayInfoText,
  learnLanguageAlpha3,
  onFinish,
  onStart,
  trainer,
  soundService
}) => {
  const initItems = useMemo(() => {
    const items = flattenItems(trainer);
    if (trainer.description) {
      items.unshift(createDescriptionAsTask(trainer.description));
    }
    return items;
  }, [trainer]);

  const {
    element: currentItem,
    elements: items,
    index: itemIndex,
    next: nextItem,
    isLast
  } = useSequence(initItems, initialItemIndex);
  const waitForDictatePlayback = useRef(Promise.resolve());

  const [playItemSound, isItemSoundPlaying] = useTrainerItemSounds(items);
  const { stop: stopAllSounds } = soundService;

  const [state, setState] = useState(TRAINER_STATE.READY);
  const showTranslation = trainer.translationVisibility === 'full';
  const [translationVisible, setTranslationVisible] = useState(showTranslation);

  useEffect(() => {
    onStart({ scorableItemsCount: items.filter(isInteractive).length });
  }, [onStart, items]);

  const handleNextItem = useCallback(() => {
    if (isLast) {
      setState(TRAINER_STATE.AWAITING_CONTINUE);
    } else {
      nextItem();
      setState(TRAINER_STATE.READY);
    }
  }, [isLast, nextItem]);

  useEffect(() => {
    if (state === TRAINER_STATE.READY) {
      // if item is non-interactive, it can be either a task or a phrase-without-gap
      // and it can have an info text
      if (!isInteractive(currentItem)) {
        setState(TRAINER_STATE.DONE);
        playItemSound(currentItem)
          .catch(() => {})
          .then(() => {
            if (currentItem?.type === 'task') {
              // Give the user some time to read the contextual info
              return wait(TIMEOUT_AFTER_CONTEXTUAL_INFO);
            }
          })
          .then(() => {
            if (currentItem?.infoText) {
              // Then show them the infotext
              displayInfoText(currentItem);
              // And give them some time to read the info text
              return wait(TIMEOUT_AFTER_INFOTEXT_DISPLAY);
            }
          })
          .finally(handleNextItem);
      } else if (trainer.dictate) {
        waitForDictatePlayback.current = playItemSound(currentItem).catch(() => {});
      }
    }
  }, [trainer, currentItem, state, displayInfoText, playItemSound, handleNextItem]);

  const handleFinish = () => {
    setState(TRAINER_STATE.DONE);
    stopAllSounds();
    onFinish();
  };

  const userInitiatedItemSound = (item) => {
    if (trainer.dictate && item === currentItem) {
      // User initiated item sound playback should block progression to the next item
      waitForDictatePlayback.current = playItemSound(item).catch(() => {});
    } else {
      playItemSound(item).catch(() => {});
    }
  };

  const handleItemComplete = useCallback(async (mistakeCount) => {
    if (currentItem.nonInteractive) return;

    if (typeof mistakeCount !== 'undefined') completeItem(currentItem, mistakeCount);

    setState(TRAINER_STATE.DONE);

    const waitForSound = trainer.dictate
      ? waitForDictatePlayback.current
      : playItemSound(currentItem).catch(() => {});

    await waitForSound
      .then(() => {
        if (currentItem.infoText) {
          displayInfoText(currentItem);
          return wait(TIMEOUT_AFTER_INFOTEXT_DISPLAY);
        }
      })
      .finally(handleNextItem);
  }, [trainer, currentItem, completeItem, handleNextItem, playItemSound, displayInfoText]);

  const handleAttempt = useCallback(({ attempt }) => {
    const trainerData = {
      /* eslint-disable camelcase */
      interaction_mode: trainer.interaction,
      item_position_in_trainer: itemIndex,
      number_of_items_in_trainer: items.length,
      translation_mode: trainer.translationVisibility
      /* eslint-enable camelcase */
    };

    const trackingAttempt = {
      ...attempt,
      text: stripHtml(attempt.inputText),
      targetText: stripHtml(attempt.text)
    };
    delete trackingAttempt.inputText;

    attemptItem(currentItem, trackingAttempt, trainerData);
  },
  [
    attemptItem,
    currentItem,
    itemIndex,
    items.length,
    trainer.interaction,
    trainer.translationVisibility
  ]
  );

  const handleToggleTranslationVisibility = () => {
    setTranslationVisible(!translationVisible);
  };

  const interactionType = getInteractionType(trainer);

  return (
    <div
      data-trainer-type="dialog"
      data-trainer-interaction={trainer.interaction}
      data-trainer-puzzlehelper={Boolean(trainer.puzzleHelper)}
      data-trainer-dictate={Boolean(trainer.dictate)}
    >
      <DialogTrainer
        active={state === TRAINER_STATE.READY}
        currentItem={currentItem}
        currentItemIndex={isLast ? null : itemIndex}
        visibleItems={items.slice(0, itemIndex + 1)}
        image={trainer.image}
        title={trainer.title}
        learnLanguageAlpha3={learnLanguageAlpha3}
        onAttempt={handleAttempt}
        onItemComplete={handleItemComplete}
        onToggleTranslationVisibility={handleToggleTranslationVisibility}
        isItemSoundPlaying={isItemSoundPlaying}
        playItemSound={userInitiatedItemSound}
        translationVisibility={trainer.translationVisibility}
        translationVisible={translationVisible}
        interaction={interactionType}
      />
      {state === TRAINER_STATE.AWAITING_CONTINUE && <ContinueSheet onClick={handleFinish} />}
    </div>
  );
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3,
  locale: session.locale
});

const mapDispatchToProps = {
  skipItem: sequenceActions.skipItem,
  attemptItem: sequenceActions.attemptItem,
  completeItem: sequenceActions.completeItem
};

DialogTrainerContainer.propTypes = {
  attemptItem: PropTypes.func.isRequired,
  currentTrainerItemIndex: PropTypes.number.isRequired,
  displayInfoText: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  onFinish: PropTypes.func.isRequired,
  completeItem: PropTypes.func.isRequired,
  onStart: PropTypes.func.isRequired,
  trainer: PropTypes.object.isRequired,
  soundService: PropTypes.any.isRequired
};

export default compose(
  withInfoText,
  withServices(['soundService']),
  connect(mapStateToProps, mapDispatchToProps),
  withNormalisedTrainerData
)(DialogTrainerContainer);
