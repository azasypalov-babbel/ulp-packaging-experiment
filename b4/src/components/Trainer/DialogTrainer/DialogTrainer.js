import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { markupStringToHTML, getFirstCorrectSolution } from '@lessonnine/babbel-markup-helper.js';
import rollbar from '../../../services/rollbarService';

import { useTheme } from '../../../providers/theme.js';
import styled from 'styled-components';
import { rem } from 'polished';
import MessageBlock from './MessageBlock';
import ContextualInfo from '../../shared/ContextualInfo';

import withTranslations from '../../shared/withTranslations';
import StyledTrainerLayout from '../../TrainerLayout';
import TranslationVisibilityToggle from '../shared/TranslationVisibilityToggle';
import TrainerTitle from '../shared/TrainerTitle';

import getComponentFromInteractionType from '../../../lib/getComponentFromInteractionType';
import usePositionMap from './usePositionMap';
import useItemGrouping from './useItemGrouping';
import Text from '../../shared/Text';
import { isInteractive, isTask } from '../../shared/itemSelectors';
import { useBackgroundColor } from '../../shared/hooks/useBackgroundColor';


const StyledWrapper = styled.div`
  margin: 0 auto;
  max-width: ${rem('700px')};
  padding: 0 ${rem('16px')};
  @media (min-width: ${(props) => props.theme.viewports.breakpoints.xxsmall}) {
    padding: 0 ${rem('24px')};
  }
  & > * + * {
    margin-top: ${rem('12px')};
    scroll-padding: ${rem('12px')};
  }
`;

const StyledContextualInfoWrapper = styled.div`
  &:not(:first-of-type) {
    margin-top: ${rem('24px')};
  }
`;

export const DialogTrainer = ({
  active,
  visibleItems: items,
  currentItem,
  learnLanguageAlpha3,
  onAttempt,
  onItemComplete,
  onToggleTranslationVisibility,
  playItemSound,
  isItemSoundPlaying,
  title,
  interaction,
  translations,
  translationVisibility,
  translationVisible
}) => {
  useEffect(() => {
    if (!currentItem.learnLanguageText && !currentItem.displayLanguageText) {
      rollbar.warning('Dialog Trainer: item with no text', { item: currentItem });
    }
  }, [currentItem]);

  const theme = useTheme();
  useBackgroundColor(theme.cascada.iceGrayW50);

  const Interaction = getComponentFromInteractionType(interaction);

  const { isFirstInGroup, getItemsInGroup } = useItemGrouping(items);
  const [getPosition] = usePositionMap(items);

  const renderRowContent = (item) => {
    const isActiveItem = active && item === currentItem;
    if (isTask(item)) {
      if (!item.displayLanguageText) return null;

      return (
        <StyledContextualInfoWrapper key={item.id}>
          <ContextualInfo
            active={isActiveItem}
            dataItemType={item.type}
            markupString={item.displayLanguageText} />
        </StyledContextualInfoWrapper>
      );
    } else {
      const hideItemWhileActive = isInteractive(item) && isActiveItem && interaction === 'speak';
      return (
        <MessageBlock
          key={item.id}
          dataItemType={item.type}
          hasAvatar={isFirstInGroup(item)}
          avatarIsSpeaking={getItemsInGroup(item).some(isItemSoundPlaying)}
          isSpeaking={isItemSoundPlaying(item)}
          speakerRole={item.speakerRole}
          secondaryText={!hideItemWhileActive && translationVisible ? item.displayLanguageText : undefined}
          position={getPosition(item)}
          onClick={item.sound && !items.some(isItemSoundPlaying) ? () => playItemSound(item) : undefined}
          active={isActiveItem}
        >
          {isInteractive(item) ?
            <div data-selector="interactive">
              <Interaction
                active={isActiveItem}
                item={item}
                onFinish={onItemComplete}
                onAttempt={onAttempt}
                learnLanguageAlpha3={learnLanguageAlpha3}
              />
            </div>
            :
            <Text
              color="spaceGray"
              data-selector="non-interactive"
              dangerouslySetInnerHTML={{
                __html: markupStringToHTML(getFirstCorrectSolution(item.learnLanguageText || ''))
              }}
            />
          }
        </MessageBlock>
      );
    }
  };

  return (
    <>
      <TranslationVisibilityToggle
        translationVisibility={translationVisibility}
        onClick={onToggleTranslationVisibility}
      />
      <StyledTrainerLayout>
        <TrainerTitle text={title || translations.trainerTitle[interaction]} />
        <StyledWrapper data-selector={`dialog-${interaction}`}>
          {items.map(renderRowContent)}
        </StyledWrapper>
      </StyledTrainerLayout>
    </>
  );
};

DialogTrainer.propTypes = {
  active: PropTypes.bool.isRequired,
  currentItem: PropTypes.object.isRequired,
  translationVisibility: PropTypes.string.isRequired,
  translationVisible: PropTypes.bool.isRequired,
  translations: PropTypes.shape({
    trainerTitle: PropTypes.shape({
      fillin: PropTypes.string.isRequired,
      puzzlehelper: PropTypes.string.isRequired,
      choose: PropTypes.string.isRequired
    })
  }).isRequired,
  title: PropTypes.string,
  visibleItems: PropTypes.array,
  interaction: PropTypes.string.isRequired,
  onAttempt: PropTypes.func.isRequired,
  onItemComplete: PropTypes.func.isRequired,
  onToggleTranslationVisibility: PropTypes.func.isRequired,
  isItemSoundPlaying: PropTypes.func.isRequired,
  playItemSound: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired
};

const getTranslations = (translate) => {
  return {
    trainerTitle: {
      fillin: translate('b3.page_dialog.complete_the_dialogue'),
      puzzlehelper: translate('b3.page_dialog.complete_the_dialogue'),
      choose: translate('b3.page_dialog.complete_the_dialogue_choicebuttons'),
      speak: translate('b3.page_dialog.title_speak_phase1')
    }
  };
};

export default withTranslations(getTranslations)(DialogTrainer);
