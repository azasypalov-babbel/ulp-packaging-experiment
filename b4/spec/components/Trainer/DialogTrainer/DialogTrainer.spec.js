import React from 'react';
import { shallow } from 'enzyme';
import { defaultPropsDialog as defaultProps } from './defaultProps';

import * as ThemeContext from '../../../../src/providers/theme.js';
import { DialogTrainer } from '../../../../src/components/Trainer/DialogTrainer/DialogTrainer';

import getComponentFromInteractionType from '../../../../src/lib/getComponentFromInteractionType';

jest.mock('../../../../src/lib/matchingUtils/normalise');
jest.mock('../../../../src/lib/getComponentFromInteractionType', () => jest.fn());

const MockInteraction = () => null;
getComponentFromInteractionType.mockImplementation(() => MockInteraction);

const mockTheme = { cascada: { iceGrayW50: 'color' } };
jest.spyOn(ThemeContext, 'useTheme').mockImplementation(() => mockTheme);

describe('<DialogTrainer />', () => {
  describe('renders', () => {
    test('with full props a list of items', () => {
      const wrapper = shallow(<DialogTrainer {...defaultProps} />);
      expect(wrapper).toMatchSnapshot();
    });

    describe('task row', () => {
      it('passes the babbel markup to the contextual info', () => {
        const wrapper = shallow(<DialogTrainer {...defaultProps} />);
        const secondTask = wrapper.find('ContextualInfo').at(1);
        const text = secondTask.prop('markupString');

        expect(text).toEqual('"Y sho era" el **mejor**');
      });

      describe('with no data (empty)', () => {
        it('returns null', () => {
          const wrapper = shallow(<DialogTrainer {...defaultProps} />);
          // demo item 7
          const emptyRow = wrapper.find('ContextualInfo').at(2);

          expect(emptyRow.exists()).toEqual(false);
        });
      });
    });

    describe('interactive phrase row', () => {
      it('passes the babbel markup to the MockInteraction', () => {
        const wrapper = shallow(<DialogTrainer {...defaultProps} />);
        const firstRow = wrapper.find('MessageBlock').at(0);
        const { learnLanguageText } = firstRow
          .find(MockInteraction)
          .prop('item');

        // eslint-disable-next-line max-len
        expect(learnLanguageText).toEqual('Ich habe ((*mich|*totally)) auch angemeldet und gehe "noch heute" zu meinem ersten Treffen!');
      });
    });

    describe('non interactive phrase row', () => {
      it('formats the item text', () => {
        const wrapper = shallow(<DialogTrainer {...defaultProps} />);
        const thirdRow = wrapper.find('MessageBlock').at(1);
        const text = thirdRow
          .find('[data-selector="non-interactive"]')
          .prop('dangerouslySetInnerHTML')['__html'];

        expect(text).toEqual('Argentina <i>became</i> <b>world champion</b>!');
      });
    });
  });

  describe('item translation text', () => {
    describe('with translations set to visible', () => {
      let wrapper;
      beforeEach(() => {
        const props = {
          ...defaultProps,
          translationVisible: true
        };
        wrapper = shallow(<DialogTrainer {...props} />);
      });
      it('should render item translation text', () => {
        const translationText = wrapper.find('MessageBlock').at(0).prop('secondaryText');
        // eslint-disable-next-line max-len
        expect(translationText).toEqual('I also signed up and I\'m going on my first date "already today"! **(the solution is: \'mich\' or \'totally\')**');
      });
    });

    describe('with translations set to not visible', () => {
      let wrapper;
      beforeEach(() => {
        const props = {
          ...defaultProps,
          translationVisible: false
        };
        wrapper = shallow(<DialogTrainer {...props} />);
      });
      it('should not render item translation text', () => {
        const translationText = wrapper.find('MessageBlock').at(0).prop('secondaryText');
        // eslint-disable-next-line max-len
        expect(translationText).toEqual(undefined);
      });
    });
  });

  describe('during purge flow', () => {
    describe('when an interactive item (phrase) was solved correctly in the lesson', () => {
      let wrapper;
      beforeEach(() => {
        const props = {
          ...defaultProps,
          visibleItems: defaultProps.visibleItems.map((item, index) => ({
            ...item,
            nonInteractive: index === 0
          }))
        };
        wrapper = shallow(<DialogTrainer {...props} />);
      });
      it('renders the phrase as non interactive', () => {
        const firstRow = wrapper.find('MessageBlock').at(0);
        expect(firstRow.find(MockInteraction).exists()).toBeFalsy();
        expect(firstRow.find('[data-selector="non-interactive"]').exists()).toBeTruthy();
      });
      it('formats the phrase text and removes gaps', () => {
        const firstRow = wrapper.find('MessageBlock').at(0);
        const text = firstRow
          .find('[data-selector="non-interactive"]')
          .prop('dangerouslySetInnerHTML')['__html'];

        expect(text).toEqual('Ich habe mich auch angemeldet und gehe <i>noch heute</i> zu meinem ersten Treffen!');
      });
      it('keeps three interactive rows', () => {
        const countInteractive = wrapper.find('[data-selector="interactive"]').length;
        const countNonInteractive = wrapper.find('[data-selector="non-interactive"]').length;
        expect(countInteractive).toEqual(1);
        expect(countNonInteractive).toEqual(4);
      });
    });
    describe('renders the formatted text correctly when item had two gaps', () => {
      let wrapper;
      beforeEach(() => {
        const props = {
          ...defaultProps,
          visibleItems: defaultProps.visibleItems.map((item, index) => ({
            ...item,
            nonInteractive: index === 5
          }))
        };
        wrapper = shallow(<DialogTrainer {...props} />);
      });
      it('formats the item text and removes gaps', () => {
        const firstRow = wrapper.find('MessageBlock').at(3);
        const text = firstRow
          .find('[data-selector="non-interactive"]')
          .prop('dangerouslySetInnerHTML')['__html'];

        expect(text).toEqual('<i>Eine</i> Woche besteht aus sieben Tagen, <b>oder?</b>');
      });
    });
  });
});
