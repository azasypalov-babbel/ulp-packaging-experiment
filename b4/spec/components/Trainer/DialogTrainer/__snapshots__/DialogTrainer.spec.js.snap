// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogTrainer /> renders with full props a list of items 1`] = `
<Fragment>
  <TranslationVisibilityToggle
    onClick={[MockFunction]}
    translationVisibility="full"
  />
  <styled.div>
    <TrainerTitle
      text="Dialog Fillin Title"
    />
    <styled.div
      data-selector="dialog-fillin"
    >
      <MessageBlock
        active={true}
        avatarIsSpeaking={false}
        dataItemType="phrase"
        hasAvatar={true}
        isSpeaking={false}
        key="a"
        onClick={[Function]}
        position="left"
        secondaryText="I also signed up and I'm going on my first date \\"already today\\"! **(the solution is: 'mich' or 'totally')**"
        speakerRole="m1"
      >
        <div
          data-selector="interactive"
        >
          <MockInteraction
            active={true}
            item={
              Object {
                "displayLanguageText": "I also signed up and I'm going on my first date \\"already today\\"! **(the solution is: 'mich' or 'totally')**",
                "id": "a",
                "infoText": "Testing 123",
                "learnLanguageText": "Ich habe ((*mich|*totally)) auch angemeldet und gehe \\"noch heute\\" zu meinem ersten Treffen!",
                "sound": Object {
                  "id": "aa",
                },
                "speakerRole": "m1",
                "type": "phrase",
              }
            }
            learnLanguageAlpha3="DEU"
            onAttempt={[MockFunction]}
            onFinish={[MockFunction]}
          />
        </div>
      </MessageBlock>
      <styled.div
        key="b"
      >
        <ContextualInfo
          active={false}
          dataItemType="task"
          markupString="Seine Freundin Aerdane ist erstaunt!"
        />
      </styled.div>
      <MessageBlock
        active={false}
        avatarIsSpeaking={false}
        dataItemType="phrase"
        hasAvatar={true}
        isSpeaking={false}
        key="c"
        onClick={[Function]}
        position="right"
        secondaryText="Argentina \\"salio\\" **campeon**"
        speakerRole="f1"
      >
        <styled.span
          color="spaceGray"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Argentina <i>became</i> <b>world champion</b>!",
            }
          }
          data-selector="non-interactive"
        />
      </MessageBlock>
      <styled.div
        key="d"
      >
        <ContextualInfo
          active={false}
          dataItemType="task"
          markupString="\\"Y sho era\\" el **mejor**"
        />
      </styled.div>
      <MessageBlock
        active={false}
        avatarIsSpeaking={false}
        dataItemType="phrase"
        hasAvatar={true}
        isSpeaking={false}
        key="e"
        onClick={[Function]}
        position="right"
        secondaryText="Pero en realidad no \\"me\\" **gusta** el fÃºtbol"
        speakerRole="f1"
      >
        <styled.span
          color="spaceGray"
          dangerouslySetInnerHTML={
            Object {
              "__html": "But <i>I</i> actually don't <b>like</b> football",
            }
          }
          data-selector="non-interactive"
        />
      </MessageBlock>
      <MessageBlock
        active={false}
        avatarIsSpeaking={false}
        dataItemType="phrase"
        hasAvatar={true}
        isSpeaking={false}
        key="f"
        onClick={[Function]}
        position="left"
        secondaryText="\\"One\\" week consists of seven days, **or not?**"
        speakerRole="m1"
      >
        <div
          data-selector="interactive"
        >
          <MockInteraction
            active={false}
            item={
              Object {
                "displayLanguageText": "\\"One\\" week consists of seven days, **or not?**",
                "id": "f",
                "learnLanguageText": "\\"Eine\\" ((Woche)) besteht aus ((sieben Tagen)), **oder?**",
                "sound": Object {
                  "id": "ff",
                },
                "speakerRole": "m1",
                "type": "phrase",
              }
            }
            learnLanguageAlpha3="DEU"
            onAttempt={[MockFunction]}
            onFinish={[MockFunction]}
          />
        </div>
      </MessageBlock>
      <MessageBlock
        active={false}
        avatarIsSpeaking={false}
        dataItemType="phrase"
        hasAvatar={true}
        isSpeaking={false}
        key="h"
        onClick={[Function]}
        position="right"
        secondaryText={null}
        speakerRole="f1"
      >
        <styled.span
          color="spaceGray"
          dangerouslySetInnerHTML={
            Object {
              "__html": "",
            }
          }
          data-selector="non-interactive"
        />
      </MessageBlock>
    </styled.div>
  </styled.div>
</Fragment>
`;
